plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java

```diff
@@ -202,6 +212,73 @@ private static boolean isAggregate(ExpressionContext expressionContext)
         return expressionContext.getType() == ExpressionContext.Type.FUNCTION && expressionContext.getFunction().getType() == FunctionContext.Type.AGGREGATION;
     }
 
+    private static boolean hasAggregate(ExpressionContext expressionContext)
+    {
+        switch (expressionContext.getType()) {
+            case IDENTIFIER:
+            case LITERAL:
+                return false;
+            case FUNCTION:
+                if (isAggregate(expressionContext)) {
+                    return true;
+                }
+                for (ExpressionContext argument : expressionContext.getFunction().getArguments()) {
+                    if (hasAggregate(argument)) {
+                        return true;
+                    }
+                }
+                return false;
+        }
+        throw new PinotException(PINOT_EXCEPTION, Optional.empty(), format("Unsupported expression type '%s'", expressionContext.getType()));
+    }
+
+    // Extracted from org.apache.pinot.core.query.reduce.AggregationDataTableReducer
+    private static DataSchema getPreAggregationDataSchema(QueryContext queryContext)
```

Comment:
nit: Move after `getAggregateTypes` since that's where it's used?

Reply:
First one is done. For the second suggestion - pinot returns arrays, do you think using Intstream.range would make it more succinct? I'll wait for your feedback.

Reply:
First one is done. For the second suggestion - pinot returns arrays, do you think using Intstream.range would make it more succinct? I'll wait for your feedback.

Reply:
nevermind for the 2nd one. It seems the DataSchema takes arrays so converting back from stream to array world seems to wipe any readability gained IMO.

I came up with:

```java
        Stream<AggregationFunction> aggregationFunctions = Arrays.stream(queryContext.getAggregationFunctions());
        String[] columnNames = aggregationFunctions
                .map(AggregationFunction::getResultColumnName)
                .collect(toImmutableList())
                .toArray(new String[0]);
        DataSchema.ColumnDataType[] columnDataTypes = aggregationFunctions
                .map(AggregationFunction::getFinalResultColumnType)
                .collect(toImmutableList())
                .toArray(new DataSchema.ColumnDataType[0]);
        return new DataSchema(columnNames, columnDataTypes);
```

Not much better than what we already have.